lib/linkerd2-proxy-resolve/src/map_endpoint.rs

```diff
@@ -0,0 +1,134 @@
+//! A middleware that wraps `Resolutions`, modifying their endpoint type.
+
+use futures::{try_ready, Async, Future, Poll};
+use linkerd2_proxy_core::resolve;
+use std::net::SocketAddr;
+
+pub trait MapEndpoint<Target, In> {
+    type Out;
+    fn map_endpoint(&self, target: &Target, addr: SocketAddr, in_ep: In) -> Self::Out;
+}
+
+#[derive(Clone, Debug)]
+pub struct Resolve<M, R> {
+    resolve: R,
+    map: M,
+}
+
+#[derive(Debug)]
+pub struct ResolveFuture<T, F, M> {
+    future: F,
+    target: Option<T>,
+    map: Option<M>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Resolution<T, M, R> {
+    resolution: R,
+    target: T,
+    map: M,
+}
+
+// === impl Resolve ===
+
+impl<M, R> Resolve<M, R> {
+    pub fn new<T>(map: M, resolve: R) -> Self
+    where
+        Self: resolve::Resolve<T>,
+    {
+        Self { resolve, map }
+    }
+}
+
+impl<T, M, R> tower::Service<T> for Resolve<M, R>
+where
+    T: Clone,
+    R: resolve::Resolve<T>,
+    M: MapEndpoint<T, R::Endpoint> + Clone,
+{
+    type Response = Resolution<T, M, R::Resolution>;
+    type Error = R::Error;
+    type Future = ResolveFuture<T, R::Future, M>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
```

Comment:
Should this be `#[inline]` to avoid pushing a stack frame for the outer `poll_ready`? IIRC rust won't inline functions across crate boundaries unless they're marked with the `#[inline]` attribute...